### PR TITLE
Log rotation fix

### PR DESCRIPTION
**Pull request by marta-lokhova**
Updates to how we perform log rotation. This change should allow for proper usage of `logrotate` command on Linux systems. The fix locks all loggers and reconfigures them which causes files to be closed and re-opened (truncated). 

I tested it by running stellar-core with a default/templated log file, then calling `logrotate`. Additionally, I tried logging from the worker thread while performing `logrotate` from the main thread.

### CODE REVIEW

**Comment by marta-lokhova** [comment]
Note that I added a lock to `registeredLoggers`; Valgrind detected that acquiring lock on Logger object and calling `getLogger` (or any function that dispatches a logger) can be potentially dangerous, as one thread could hold a lock to the logger object, while the other thread had a lock on registeredLoggers. Looks like with this change, valgrind does not produce anymore deadlock warnings when calling `logrotate`.

File: src/util/Logging.h [hunk 1]
```diff
@@ -13,22 +13,39 @@
...
+        T* const mItem;
+        LockElObject(T* elObj);
+        ~LockElObject();
+
+        // Disable copy construction and copy assignment
+        LockElObject(LockElObject const&) = delete;
+        LockElObject& operator=(LockElObject const&) = delete;
+    };
+
     struct LockHelper
```
Comment (jonjove) [inline src/util/Logging.h]: `LockElObject` and `LockHelper` are only used in the `.cpp` file so they should be moved there. Cleaner and will make compilation faster if future maintainers ever change them.

**Comment by marta-lokhova** [comment]
@jonjove updated (also squashed some commits)

**Comment by jonjove** [comment]
r+ 6cd956b

**Pull request merged**